README.md
Copyright 2014-2015 NZRS Ltd

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program. If not, see http://www.gnu.org/licenses/.

# WaveTrace #

This application allows for batch processing of radio propagation modelling.  It takes a CSV as input and generates the following:

1. A KML file with a georeferenced image for use in the likes of Google Earth.
2. A raster file suitable for use in GIS packages.
3. A shapefile (currently not implemented but trivial).

The model is derived from a Digital Elevation Model (DEM), the base data at present is from the NASA Shuttle Radio Topography Mission (SRTM).  Though there may be bettter localised sources available.

## Set Up ##

Install the following packages [setup.md](https://github.com/NZRegistryServices/wavetrace/blob/master/setup.md)

## Parameters Used in Modelling Coverage ##

* network_name	
* site_name	
* latitude	
* longitude	
* antenna height	
* frequency (MHz)	
* power (EIRP)	
* polarisation	
* bearing	
* horizontal beamwidth	
* vertical beamwidth	
* antenna downtilt

The [Longely Rice Prediction Model](http://en.wikipedia.org/wiki/Longley%E2%80%93Rice_model "Link to Wikipedia") is used to model coverage.

## The Process of Creating Coverage - Short Version##

1. Get the Digital Elevation Model Files and convert to .sdf 
  * Either get from NASA using get_data.py which scrapes and converts to .sdf files, or;
  * Copy already processed .sdf files from a local source
  * Or download from NZRS (once we have a site up)
  * Pass -h to enable splat-hd when using get_data.py
2. Ensure you have a .csv file that matches the column headings of  sample_data.csv in this repository.  The file must contain the following:
  * network_name	
  * site_name	latitude	
  * longitude	
  * antenna_height	
  * frequency_mhz	
  * power_eirp
  
  It may optionally contain

  * polarisation	
  * bearing	
  * horizontal_beamwidth	
  * vertical_beamwidth
  * Antenna downtilt
3. Use make_files.py to create the model files based off of the .csv file.
  * Azimuth file
  * Elevation file 
  * QTH file
  * LRP file
4. Map coverage using create_output_from_dir.py.  The following will be created:
  * A KML file you can use in Google Earth
  * A georeferenced .tif file (raster - 1 file)
  * A shapefile (vector - 4 files)

## The Process of Creating Coverage - Long Version##

## Acquiring and Processing DEM Data ##

The get_data.py scrapes the data from NASA.  Some tweaks can be made to limit the area by specifying the latitude and longitude.  At present the default is to acquire New Zealand data only.  Though the query_url in 'get_data.py' can be modified easily.

get_data.py also processes the data into what can be used in the model.  It unzips and converts to .sdf and cleans up intermediate files.

### Usage: ###
  
  `python get_data.py`

If using splat-hd, to get srtm1 data (1 Arc Second), add the HD flag to scrape the srtm1 data:
  
  `python get_data.py -h`
  
get_data.py does the following:

* Scrapes the Shuttle Radar Topography Mission (SRTM) data from NASA
* Unzips the downloaded data
* Converts the downloaded .hgt files to .sdf files
* Cleans stuff up

  
## Creating Files For the Model ##
The model uses four files, two required and two optional but created empty by default.
Required files:
* LRP - Contains Longley Rice Parameters. Where not read from the .csv file they default to New Zealand specific parameters, they may need changed for other climates and locales.
  * Earth Dielectric Constant (Relative permittivity)
  * Earth Conductivity (Siemens per meter)
  * Atmospheric Bending Constant (N-units)
  * Frequency in MHz (20 MHz to 20 GHz)
  * Climate type (Maritime Temperate, over land)
  * Polarization (0 = Horizontal, 1 = Vertical)
  * Fraction of situations (50% of locations)
  * Fraction of time (50% of the time)
  * Power EIRP
* QTH - Geographic
  * Latitude
  * Longitude
  * Antenna height above ground
* AZ - Azimuth
  * Bearing
  * Horizontal beamwidth
* EL - Elevation
  * Downtilt
  * Vertical beamwidth

The AZ and EL files use a simplified model where 90% of the energy is transmitted in the specified beamwidth and the other 10% outside of the beamwidth.  Both files can hold full antenna patterns but this has been simplified for batch modelling.

### Usage ###

 ``python make_files.py``
  
## Create GIS Outputs ##
``python create_output_from_dir.py``  

Pass -h to run splat-hd (note: this requires srtm1 data - use `python get_data.py -h`)

``python create_output_from_dir.py -h``

## Mapping coverage ##
This requires all the following files to be in the same directory:
 * The .sdf files
 * LRP files
 * QTH files
 * EZ files
 * EL files

The above files will be processed when create_ouput_from_dir.py is run.  The default system receive is modelled by default at -110 dBm receive.  Other receives can be passed in as a parameter.   Note this is at the chip, so includes receive antenna gain.

You may need to change received values to get something that feels right.

### Usage ###
`python create_output_from_dir.py`

or for say -105dBm receive

Receive_sensitivity can be controlled with -s, or --sensitivity
Positive receive_sensitivity numbers are forced to negative

``./create_output_from_folder.py -h -s -105``
``./create_output_from_folder.py -h --sensitivity=105``
``./create_output_from_folder.py -hs -105``
``./create_output_from_folder.py -s 105```


## Converting GeoTiffs to Shapefile ##

Currently this is not implemented but at the command line:

For all GeoTiffs in the folder:

``for f in *.tif;  do gdal_polygonize.py $f -f "ESRI Shapefile" $f.shp;  done```

For a single GeoTiff:

``gdal_polygonize.py <geotiff name> -f "ESRI Shapefile" <output_file_name.shp>```










CHANGELOG
Changelog
==========

v4.0.3, 2018-10-31
-------------------
- Updated support for GitLab API v4.
- Updated Python libraries.


v4.0.2, 2017-05-08
-------------------
- Updated URL for ``main.get_geoid_height()`


v4.0.1, 2017-04-21
-------------------
- Minor changes.


v4.0.0, 2017-01-17
-------------------
- Removed Gitlab API key option, because it is unnecessary for accessing the required public repository
- Added the option to create ESRI Shapefiles in the functions ``main.compute_coverage`` and ``main.compute_satellite_los``.


v3.0.1, 2016-12-15
-------------------
- Repackaged with setuptools and ensured that the README is included in tar archive this time


v3.0.0, 2016-08-22
-------------------
- Added functions to compute line-of-sight for geostationary satellites
- Renamed some functions, hence the version bump


v2.0.0, 2016-08-22
-------------------
- Complete refactor and conversion into a Python package


v1.0.0, 2013
--------------
- Initial version 


wavetrace/constants.py
import os
from pathlib import Path


PROJECT_ROOT = Path(os.path.abspath(os.path.join(
  os.path.dirname(__file__), '../')))
SECRETS_PATH = PROJECT_ROOT/'secrets.json'

#: SRTM tiles that cover New Zealand; for a visual, see http://geojson.io/#id=gist:anonymous/81b4cb465f1c78941f665c9038494f0f&map=5/-41.360/172.463
SRTM_NZ_TILE_IDS = [
  'S35E172',
  'S35E173',
  'S36E173',
  'S36E174',
  'S36E175',
  'S37E173',
  'S37E174',
  'S37E175',
  'S37E176',
  'S38E174',
  'S38E175',
  'S38E176',
  'S38E177',
  'S38E178',
  'S39E174',
  'S39E175',
  'S39E176',
  'S39E177',
  'S39E178',
  'S40E173',
  'S40E174',
  'S40E175',
  'S40E176',
  'S40E177',
  'S40E178',
  'S41E172',
  'S41E173',
  'S41E174',
  'S41E175',
  'S41E176',
  'S42E171',
  'S42E172',
  'S42E173',
  'S42E174',
  'S42E175',
  'S42E176',
  'S43E170',
  'S43E171',
  'S43E172',
  'S43E173',
  'S43E174',
  'S44E168',
  'S44E169',
  'S44E170',
  'S44E171',
  'S44E172',
  'S44E173',
  'S45E167',
  'S45E168',
  'S45E169',
  'S45E170',
  'S45E171',
  'S46E166',
  'S46E167',
  'S46E168',
  'S46E169',
  'S46E170',
  'S46E171',
  'S47E166',
  'S47E167',
  'S47E168',
  'S47E169',
  'S47E170',
  'S48E167',
  'S48E168',
  ]

#: Transmitter CSV files must have these header columns
REQUIRED_TRANSMITTER_FIELDS = [
  'network_name',    
  'site_name',
  'latitude', # WGS84 float
  'longitude', # WGS84 float 
  'antenna_height', # meters
  'polarization', # 0 (horizontal) or 1 (vertical)
  'frequency', # megaherz
  'power_eirp', # watts
  ]

#: SPLAT! Earth dielectric constant.
#: According to the SPLAT! documentation, typical Earth dielectric constants and conductivities are: 
#: Salt water, 80, 5.000;
#: Good ground, 25, 0.020;
#: Fresh water, 80, 0.010;
#: Marshy land, 12, 0.007;
#: Farmland or forest, 15, 0.005;
#: Average ground, 15, 0.005;
#: Mountain or sand, 13, 0.002;
#: City, 5, 0.001;
#: Poor ground, 4, 0.001;
EARTH_DIELECTRIC_CONSTANT = 15

#: SPLAT! Earth earth_conductivity in Siemens per meter
EARTH_CONDUCTIVITY = 0.005

#: SPLAT! radio climate codes.
#: 1=Equatorial (Congo);
#: 2=Continental Subtropical (Sudan);
#: 3=Maritime Subtropical (West coast of Africa);
#: 4=Desert (Sahara);
#: 5=Continental Temperate;
#: 6=Maritime Temperate, over land (UK and west coasts of US & EU);
#: 7=Maritime Temperate, over sea
RADIO_CLIMATE = 6

#: SPLAT! time variability parameter
FRACTION_OF_TIME = 0.5

#: SPLAT! location variability parameter
FRACTION_OF_SITUATIONS = 0.5

#: SPLAT receiver sensitivity parameter in decibel-milliwatts (dBm). 
#: For example, minimum received signal power of wireless networks (802.11 variants) is -100 dBm.
RECEIVER_SENSITIVITY = -110 
#: WGS84 semimajor axis in meters
WGS84_A = 6378137
#: WGS84 flattening 
WGS84_F = 1/298.257223563
#: WGS84 eccentricity squared (e^2)
WGS84_E2 = 2*WGS84_F - WGS84_F**2
#: Distance in meters of a geostationary satellite from the center of the Earth (and hence the center of the WGS84 ellipsoid);
#: taken from the Wikipedia article `Geostationary orbit <https://en.wikipedia.org/wiki/Geostationary_orbit>`_
R_S = 42164000
#: Distance in meters of a geostationary satellite from the WGS84 ellipsoid
H_S = R_S - WGS84_A